Source/Fuse.Drawing.Surface/Android/GraphicsSurface.uno

```diff
@@ -0,0 +1,161 @@
+			if (fb.Size.X == 0 || fb.Size.Y == 0)
+			{
+				return;
+			}
+			LoadBitmap(impl, fb.Size.X, fb.Size.Y);
+			BeginImpl(impl, fb.Size.X, fb.Size.Y, (int)fb.ColorBuffer.GLTextureHandle);
+		}
+
+		/**
+			Load a bitmap of given dimensions into the context and use it for the canvas
+		*/
+		[Foreign(Language.Java)]
+		public static extern(Android) void LoadBitmap(Java.Object context, int width, int height)
+		@{
+			GraphicsSurfaceContext impl = (GraphicsSurfaceContext) context;
+			Bitmap b = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
+
+			Canvas canvas = impl.canvas;
+			canvas.setBitmap(b);
+			impl.bitmap = b;
+
+			canvas.setMatrix(null);
+
+			// invert our bitmap since the Android canvas is inversed when drawing
+			canvas.translate(0.0f, (float)height);
+			canvas.scale(1, -1);
+		@}
+
+		[Foreign(Language.Java)]
+		static void BeginImpl(Java.Object _context, int width, int height, int glTextureId)
+		@{
+			GraphicsSurfaceContext context = (GraphicsSurfaceContext) _context;
+			context.width = width;
+			context.height = height;
+			context.glTextureId = glTextureId;
+		@}
+
+		/**
+			Ends drawing. All drawing called after `Begin` and to now must be completed by now. This copies the resulting image to the desired output setup in `Begin`.
+		*/
+		public override void End()
+		{
+			var impl = SurfaceContext;
+
+			if (impl == null) return;
+			EndImpl(impl);
+		}
+
+		[Foreign(Language.Java)]
+		static void EndImpl(Java.Object context)
+		@{
+			GraphicsSurfaceContext realContext = (GraphicsSurfaceContext) context;
+
+			GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, realContext.glTextureId);
+
+			// heat up the caches. not needed but good to have
+			realContext.bitmap.prepareToDraw();
+
+			GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, realContext.bitmap, 0);
+			realContext.bitmap.recycle();
+		@}
+
+		/*
+			This approach is really bad now. When Erik refactors ImageSource we shouldn't
+			need to do the round-trip to GL.
+			We might end up not supporting ImageFill until this is fixed, but this is useful
+			here now to complete/test the sizing/tiling support.
+		*/
+		protected sealed override Java.Object PrepareImageFillImpl( ImageFill img )
+		{
+			var src = img.Source;
+			var tex = src.GetTexture();
+			var fb = FramebufferPool.Lock( src.PixelSize, Uno.Graphics.Format.RGBA8888, false );
+
+			_drawContext.PushRenderTarget(fb);
+			AndroidGraphicsDrawHelper.Singleton.DrawImageFill(tex);
+			Java.Object imageRef = LoadImage((int)tex.GLTextureHandle, src.PixelSize.X, src.PixelSize.Y );
+			FramebufferPool.Release(fb);
+			_drawContext.PopRenderTarget();
+
+			return imageRef;
+		}
+
+		[Foreign(Language.Java)]
+		static Java.Object LoadImage(int glTextureId, int width, int height)
+		@{
+			int size = width * height * 4;
+			int[] pixels = new int[size];
+
+			IntBuffer pixelData = IntBuffer.wrap(pixels);
+			GLES20.glReadPixels(0, 0, width,height, GLES20.GL_RGBA, GLES20.GL_UNSIGNED_BYTE, pixelData);
```

Comment:
You need to do `GLES20.glPixelStorei(GLES20.GL_UNPACK_ALIGNMENT, 1);` before this, otherwise you risk writing past the end of `pixels` (and also getting garbage) on some image sizes. We already have this problems other places, but please don't repeat the mistake ;)